File: apps/ios/GalaxySSI/AgentResponseSections.swift
import Foundation

enum AgentResponseSectionKind: String, Codable, CaseIterable, Identifiable {
  case plan = "PLAN"
  case executionLog = "EXECUTION_LOG"
  case finalAnswer = "FINAL_ANSWER"
  case evidence = "EVIDENCE"

  var id: String { rawValue }
}

struct AgentResponseSection: Codable, Equatable, Identifiable {
  var kind: AgentResponseSectionKind
  var blocks: [AgentRichBlock]
  var expandedByDefault: Bool

  var id: String { kind.rawValue }
}

struct AgentResponseSectionLayout: Codable, Equatable {
  var collapsible: Bool
  var sections: [AgentResponseSection]
}

final class AgentResponseSectionExpansionStore {
  static let shared = AgentResponseSectionExpansionStore()

  private let defaults: UserDefaults
  private let key: String
  private var states: [String: Bool]

  init(defaults: UserDefaults = .standard, key: String = "galaxyssi.agent.response.section.expansion") {
    self.defaults = defaults
    self.key = key
    if let data = defaults.data(forKey: key),
       let decoded = try? JSONDecoder().decode([String: Bool].self, from: data) {
      states = decoded
    } else {
      states = [:]
    }
  }

  func state(for storageKey: String) -> Bool? {
    guard !storageKey.isEmpty else { return nil }
    return states[storageKey]
  }

  func set(_ expanded: Bool, for storageKey: String) {
    guard !storageKey.isEmpty else { return }
    states[storageKey] = expanded
    while states.count > Self.maximumEntries, let firstKey = states.keys.first {
      states.removeValue(forKey: firstKey)
    }
    guard let data = try? JSONEncoder().encode(states) else { return }
    defaults.set(data, forKey: key)
  }

  private static let defaultKey = "galaxyssi.agent.response.section.expansion"
  private static let maximumEntries = 512
}

enum AgentResponseSectionOrganizer {
  static func organize(
    _ blocks: [AgentRichBlock],
    expandStructuredDetails: Bool = false
  ) -> AgentResponseSectionLayout {
    guard !blocks.isEmpty else {
      return AgentResponseSectionLayout(collapsible: false, sections: [])
    }

    var grouped: [AgentResponseSectionKind: [AgentRichBlock]] = [:]
    var activeKind: AgentResponseSectionKind?
    var structured = false

    for block in blocks {
      if let heading = headingKind(block) {
        activeKind = heading
        structured = true
        continue
      }

      let explicit = explicitKind(block)
      if explicit != nil {
        structured = true
      }

      let kind = explicit ?? typeKind(block) ?? activeKind ?? .finalAnswer
      grouped[kind, default: []].append(block)
    }

    var sections = orderedKinds.compactMap { kind -> AgentResponseSection? in
      guard let sectionBlocks = grouped[kind], !sectionBlocks.isEmpty else {
        return nil
      }
      return AgentResponseSection(
        kind: kind,
        blocks: sectionBlocks,
        expandedByDefault: kind == .finalAnswer || expandStructuredDetails
      )
    }

    if !sections.contains(where: \.expandedByDefault), !sections.isEmpty {
      sections[0].expandedByDefault = true
    }

    let contentSize = blocks.reduce(0) { $0 + contentCharacters($1) }
    let longResponse = contentSize >= longResponseCharacters || blocks.count >= longResponseBlocks
    return AgentResponseSectionLayout(
      collapsible: !sections.isEmpty && (longResponse || structured),
      sections: sections
    )
  }

  private static func explicitKind(_ block: AgentRichBlock) -> AgentResponseSectionKind? {
    let value = explicitMetadataKeys
      .compactMap { block.metadata[$0] }
      .first { !$0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
    guard let value else { return nil }
    return parseKind(value)
  }

  private static func headingKind(_ block: AgentRichBlock) -> AgentResponseSectionKind? {
    guard block.type == .heading else { return nil }
    let value = block.text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? block.title : block.text
    return parseKind(value)
  }

  private static func typeKind(_ block: AgentRichBlock) -> AgentResponseSectionKind? {
    switch block.type {
    case .status, .progress, .tool, .timeline:
      return .executionLog
    case .citation:
      return .evidence
    case .actions, .approval, .form:
      return .finalAnswer
    default:
      return block.metadata["evidence"] == "true" ? .evidence : nil
    }
  }

  private static func parseKind(_ value: String) -> AgentResponseSectionKind? {
    let normalized = value
      .trimmingCharacters(in: .whitespacesAndNewlines)
      .lowercased()
      .replacingOccurrences(of: "[^\\p{L}\\p{N}]+", with: " ", options: .regularExpression)
      .trimmingCharacters(in: .whitespacesAndNewlines)

    switch normalized {
    case "plan", "approach", "steps",
      "\u{8ba1}\u{5212}", "\u{65b9}\u{6848}", "\u{6267}\u{884c}\u{8ba1}\u{5212}":
      return .plan
    case "log", "execution log", "tool log", "activity", "progress",
      "\u{65e5}\u{5fd7}", "\u{6267}\u{884c}\u{65e5}\u{5fd7}", "\u{5de5}\u{5177}\u{65e5}\u{5fd7}",
      "\u{6267}\u{884c}\u{8fc7}\u{7a0b}":
      return .executionLog
    case "final", "final answer", "answer", "result", "summary",
      "\u{6700}\u{7ec8}\u{7b54}\u{6848}", "\u{7b54}\u{6848}", "\u{7ed3}\u{679c}", "\u{603b}\u{7ed3}":
      return .finalAnswer
    case "evidence", "source", "sources", "citation", "citations", "references",
      "\u{8bc1}\u{636e}", "\u{6765}\u{6e90}", "\u{5f15}\u{7528}", "\u{53c2}\u{8003}":
      return .evidence
    default:
      return nil
    }
  }

  private static func contentCharacters(_ block: AgentRichBlock) -> Int {
    block.title.utf16.count +
      block.text.utf16.count +
      block.fallbackText.utf16.count +
      block.rows.reduce(0) { total, row in
        total + row.reduce(0) { $0 + $1.utf16.count }
      }
  }

  private static let longResponseCharacters = 1_200
  private static let longResponseBlocks = 6
  private static let explicitMetadataKeys = ["section", "response_section", "role"]
  private static let orderedKinds: [AgentResponseSectionKind] = [
    .plan,
    .executionLog,
    .finalAnswer,
    .evidence
  ]
}

File: apps/android/app/src/main/java/com/galaxyssi/chat/AgentResponseSections.kt
package com.galaxyssi.chat

import java.util.Locale

enum class AgentResponseSectionKind {
    PLAN,
    EXECUTION_LOG,
    FINAL_ANSWER,
    EVIDENCE
}

data class AgentResponseSection(
    val kind: AgentResponseSectionKind,
    val blocks: List<AgentRichBlock>,
    val expandedByDefault: Boolean
)

data class AgentResponseSectionLayout(
    val collapsible: Boolean,
    val sections: List<AgentResponseSection>
)

object AgentResponseSectionOrganizer {
    private const val LONG_RESPONSE_CHARACTERS = 1_200
    private const val LONG_RESPONSE_BLOCKS = 6

    private val orderedKinds = listOf(
        AgentResponseSectionKind.PLAN,
        AgentResponseSectionKind.EXECUTION_LOG,
        AgentResponseSectionKind.FINAL_ANSWER,
        AgentResponseSectionKind.EVIDENCE
    )

    fun organize(blocks: List<AgentRichBlock>): AgentResponseSectionLayout {
        if (blocks.isEmpty()) return AgentResponseSectionLayout(false, emptyList())
        val grouped = linkedMapOf<AgentResponseSectionKind, MutableList<AgentRichBlock>>()
        var activeKind: AgentResponseSectionKind? = null
        var structured = false

        blocks.forEach { block ->
            headingKind(block)?.let { heading ->
                activeKind = heading
                structured = true
                return@forEach
            }
            val explicit = explicitKind(block)
            if (explicit != null) structured = true
            val kind = explicit ?: typeKind(block) ?: activeKind ?: AgentResponseSectionKind.FINAL_ANSWER
            grouped.getOrPut(kind, ::mutableListOf) += block
        }

        val sections = orderedKinds.mapNotNull { kind ->
            grouped[kind]?.takeIf { it.isNotEmpty() }?.let { sectionBlocks ->
                AgentResponseSection(
                    kind = kind,
                    blocks = sectionBlocks,
                    expandedByDefault = kind == AgentResponseSectionKind.FINAL_ANSWER
                )
            }
        }.let { ordered ->
            if (ordered.none { it.expandedByDefault } && ordered.isNotEmpty()) {
                ordered.mapIndexed { index, section ->
                    if (index == 0) section.copy(expandedByDefault = true) else section
                }
            } else {
                ordered
            }
        }
        val contentSize = blocks.sumOf(::contentCharacters)
        val longResponse = contentSize >= LONG_RESPONSE_CHARACTERS ||
            blocks.size >= LONG_RESPONSE_BLOCKS
        return AgentResponseSectionLayout(
            collapsible = sections.isNotEmpty() && (longResponse || structured),
            sections = sections
        )
    }

    private fun explicitKind(block: AgentRichBlock): AgentResponseSectionKind? {
        val value = sequenceOf("section", "response_section", "role")
            .mapNotNull(block.metadata::get)
            .firstOrNull(String::isNotBlank)
            ?: return null
        return parseKind(value)
    }

    private fun headingKind(block: AgentRichBlock): AgentResponseSectionKind? {
        if (block.type != AgentRichBlockType.HEADING) return null
        return parseKind(block.text.ifBlank { block.title })
    }

    private fun typeKind(block: AgentRichBlock): AgentResponseSectionKind? = when (block.type) {
        AgentRichBlockType.STATUS,
        AgentRichBlockType.PROGRESS,
        AgentRichBlockType.TOOL,
        AgentRichBlockType.TIMELINE -> AgentResponseSectionKind.EXECUTION_LOG
        AgentRichBlockType.CITATION -> AgentResponseSectionKind.EVIDENCE
        AgentRichBlockType.ACTIONS,
        AgentRichBlockType.APPROVAL,
        AgentRichBlockType.FORM -> AgentResponseSectionKind.FINAL_ANSWER
        else -> if (block.metadata["evidence"] == "true") {
            AgentResponseSectionKind.EVIDENCE
        } else {
            null
        }
    }

    private fun parseKind(value: String): AgentResponseSectionKind? {
        val normalized = value
            .trim()
            .lowercase(Locale.ROOT)
            .replace(Regex("[^\\p{L}\\p{N}]+"), " ")
            .trim()
        return when (normalized) {
            "plan", "approach", "steps",
            "\u8ba1\u5212", "\u65b9\u6848", "\u6267\u884c\u8ba1\u5212" ->
                AgentResponseSectionKind.PLAN
            "log", "execution log", "tool log", "activity", "progress",
            "\u65e5\u5fd7", "\u6267\u884c\u65e5\u5fd7", "\u5de5\u5177\u65e5\u5fd7",
            "\u6267\u884c\u8fc7\u7a0b" -> AgentResponseSectionKind.EXECUTION_LOG
            "final", "final answer", "answer", "result", "summary",
            "\u6700\u7ec8\u7b54\u6848", "\u7b54\u6848", "\u7ed3\u679c", "\u603b\u7ed3" ->
                AgentResponseSectionKind.FINAL_ANSWER
            "evidence", "source", "sources", "citation", "citations", "references",
            "\u8bc1\u636e", "\u6765\u6e90", "\u5f15\u7528", "\u53c2\u8003" ->
                AgentResponseSectionKind.EVIDENCE
            else -> null
        }
    }

    private fun contentCharacters(block: AgentRichBlock): Int =
        block.title.length +
            block.text.length +
            block.fallbackText.length +
            block.rows.sumOf { row -> row.sumOf(String::length) }
}
